Introduce Auto-Generated Native Struct Wrappers

Kasuromi opened the pull request:
## Summary
Removes the old struct wrappers in favor of auto-generated ones for:
  - Il2CppAssembly
 - Il2CppType
 - Il2CppClass
 - Il2CppPropertyInfo
 - Il2CppMethodInfo
 - Il2CppImage
 - Il2CppFieldInfo
 - Il2CppException
 - Il2CppEventInfo

Auto generated classes follow the naming scheme: `Type_MetadataVersion_ChangeNumber`

*Il2CppParameterInfo wrappers have been left untouched as there is no need to replace them since Il2CppType is used in their place for unity version 2021.2.0+ (metadata v29+)*

Added NotSupportedExceptions to the CreateNewStruct method in Il2CppClass wrappers for unity versions below 5.5.0 as their vtable format is (and was) not supported by ClassInjector

Introduced wrappings for Il2CppAssemblyName

Il2CppException struct wrappers now use Il2CppString* instead of IntPtr to explicitly state that the string is an il2cpp string instead of a normal c-style string.

## Notes
Auto-generated wrappers include metadata v29 which is brand new to unhollower and needs testing.
Auto-generated wrappers have only been tested on unity version 2019.4.21f1

ghorsington reviewed (commented):
This looks overall good, but I still want to test it all before merging. I added one comment, but the PR itself needs some minor testing (mainly making sure nothing breaks). I'll make binaries and a test BepInEx build for quick testing.

ghorsington commented:
I also must mention that the automatic wrapper generator is great! Do you plan on releasing it as well?

ghorsington commented:
Test BepInEx builds:

* Il2Cpp x86: [UnityIL2CPP_x86.zip](https://github.com/BepInEx/Il2CppAssemblyUnhollower/files/8532592/UnityIL2CPP_x86.zip)
* Il2Cpp x64: [UnityIL2CPP_x64.zip](https://github.com/BepInEx/Il2CppAssemblyUnhollower/files/8532588/UnityIL2CPP_x64.zip)

Kasuromi commented on UnhollowerBaseLib/Runtime/VersionSpecific/MethodInfo/Interfaces.cs:
```diff
@@ -4,24 +4,13 @@ namespace UnhollowerBaseLib.Runtime.VersionSpecific.MethodInfo
-        INativeMethodInfoStruct CreateNewMethodStruct();
+        INativeMethodInfoStruct CreateNewStruct();
         unsafe INativeMethodInfoStruct Wrap(Il2CppMethodInfo* methodPointer);
-        IntPtr GetMethodFromReflection(IntPtr method);
```
Yeah it can be reverted, I'll need to add support for generating these methods as the generator I made doesn't support inserting methods into the handler class.

Kasuromi replied:
I should also probably dump a list of methods in the il2cpp api

ghorsington replied:
Seems like the specific version of Unity where I needed to manually resolve the MethodInfo is lost. It's probably better to just test and fix it later down the line when possible. As such, I'll mark this as done and merge the code in.

Kasuromi commented:
> I also must mention that the automatic wrapper generator is great! Do you plan on releasing it as well?

It's a bit of a mess and not really my proudest creation :D. I'm happy to share the source code with you if you want to check it out

ghorsington commented:
> I'm happy to share the source code with you if you want to check it out

I'm thinking more about including the generator in this repo or some other repo. The main point is that we could have the source if possible, and even automate running it in the future.

Kasuromi commented:
Oh yeah for sure, I'm happy to have it public. I was thinking you might want to look at it before we do that

Kasuromi commented:
Injection on games that use metadata v29 seems to cause a crash with these wrappers. It also crashes with the old ones, a change to ClassInjector to support the new metadata handles might be necessary.

Looking into it further at the moment

Kasuromi commented:
After looking into it, it was image metadata not being allocated that causes crashes on v27+.
Added a fix for it and implemented the struct in `Il2CppStructs.cs`. Wrappings will need to be added in the future in case unity changes them

ghorsington commented:
From a look at this, is it otherwise done? From some minor tests, it appears to work on basic games. 

If you don't plan on adding anything more into this, I think we can merge this in and start the proper refactoring process of the project.

Kasuromi commented:
It can be merged, I wasn't sure if you were going to be the one to revert GetMethodFromReflection.

In any case, these wrappers have given no issue so far so they're good to be merged

ghorsington commented:
> I wasn't sure if you were going to be the one to revert GetMethodFromReflection.

Sure, I can revert it this weekend once I get back to working on this. Sorry for the wording; I initially meant to ask if reverting it would cause any issues (as I thought there was some compatibility reason to remove it).
